tests/util/pilotServer.go

```diff
@@ -37,10 +42,59 @@ var (
 
 	// MockPilotGrpcAddr is the address to be used for grpc connections.
 	MockPilotGrpcAddr string
-	fsRoot            string
-	stop              chan struct{}
+
+	// MockPilotHttpPort is the dynamic port for pilot http
+	MockPilotHTTPPort int
+
+	// MockPilotGrpcPort is the dynamic port for pilot grpc
+	MockPilotGrpcPort int
+
+	fsRoot string
+	stop   chan struct{}
+)
+
+var (
+	// IstioTop has the top of the istio tree, matches the env variable from make.
+	IstioTop = os.Getenv("TOP")
+
+	// IstioSrc is the location if istio source ($TOP/src/istio.io/istio
+	IstioSrc = os.Getenv("ISTIO_GO")
+
+	// IstioBin is the location of the binary output directory
+	IstioBin = os.Getenv("ISTIO_BIN")
+
+	// IstioOut is the location of the output directory ($TOP/out)
+	IstioOut = os.Getenv("ISTIO_OUT")
+
+	// EnvoyOutWriter captures envoy output
+	// Redirect out and err from envoy to buffer - coverage tests get confused if we write to out.
+	// TODO: use files
+	EnvoyOutWriter bytes.Buffer
+
+	// EnvoyErrWriter captures envoy errors
+	EnvoyErrWriter bytes.Buffer
 )
 
+func init() {
+	if IstioTop == "" {
+		// Assume it is run inside istio.io/istio
+		current, _ := os.Getwd()
+		idx := strings.Index(current, "/src/istio.io/istio")
+		if idx > 0 {
+			IstioTop = current[0:idx]
+		}
+	}
+	if IstioSrc == "" {
+		IstioSrc = IstioTop + "/src/istio.io/istio"
+	}
+	if IstioOut == "" {
+		IstioOut = IstioTop + "/out"
+	}
+	if IstioBin == "" {
+		IstioBin = IstioTop + "/out/" + runtime.GOOS + "_" + runtime.GOARCH + "/release"
+	}
```

Comment:
i'd recommend failing if these are unset. letting it run by guessing values leads to difficult to understand program behavior down the line.

Reply:
It's for IDE support - improves developer experience a lot to not have to add flags/envs.